run-regression-test: support rewriting "column1 @ \"keyword\" && column2 @~ \"^(?!.*keyword1|keyword2|...).+$\""

komainu8 opened the pull request:
This feature that for backward compatibility.

A new Groonga( version 5.0.7 or later) normalizes match target text by NormalizerAuto normalizer when Groonga doesn’t use an index for regular expression search.

Therefore, there is a possibility to a different result of regular expression search between an old Groonga and a new Groonga.

This feature replaces a complex regular expression that cannot use the index to doesn't use a regular expression.

kenhys commented on test/test-filter-rewriter.rb:
```diff
@@ -110,4 +110,21 @@ def test_under_score
+                   rewrite("column1 @ \"value1\" && column2 @~ \"^(?!.*value2 | value3 | value4).+$\""))
+    end
+  end
 end
```
How about non rewrite test case?
I mean that it is  good for rewrite flag is enabled test case, but what about disabled test case?

```ruby
diff --git a/test/test-filter-rewriter.rb b/test/test-filter-rewriter.rb                                                                   
index ea76da8..f501065 100644
--- a/test/test-filter-rewriter.rb
+++ b/test/test-filter-rewriter.rb
@@ -112,9 +112,9 @@ class FilterRewriterTest < Test::Unit::TestCase
   end

   class RegularExpressionTest < self
-    def rewrite(filter)
+    def rewrite(filter, enabled=true)
       super(filter,
-            :rewrite_regular_expression => true)
+            :rewrite_regular_expression => enabled)
     end
 
     def test_rewrite_one
@@ -126,5 +126,11 @@ class FilterRewriterTest < Test::Unit::TestCase
       assert_equal("column1 @ \"value1\" &! column2 @ \"value2\" &! column2 @ \"value3\" &! column2 @ \"value4\"",
                    rewrite("column1 @ \"value1\" && column2 @~ \"^(?!.*value2 | value3 | value4).+$\""))
     end
+
+    def test_no_rewrite
+      assert_equal("column1 @ \"value1\" && column2 @~ \"^(?!.*value2).+$\"",
+                   rewrite("column1 @ \"value1\" && column2 @~ \"^(?!.*value2).+$\"", false))
+    end
+
   end
 end

```

komainu8 commented:
@kenhys Thank you for your comments.
I've fixed them.

kenhys commented:
LGTM :+1:

kou commented on lib/groonga-query-log/command/run-regression-test.rb:
```diff
@@ -187,6 +188,14 @@ def create_option_parser
                   "specifying this option multiple times") do |accessor|
           @nullable_reference_number_accessors << accessor
         end
+        parser.on("--[no-]rewrite_regular_expression",
```
Could you use `-` as word separator for consistency?

kou replied:
This name is too generic.
We don't rewrite all regular expression, right?

`--rewrite-not-or-regular-expression` or something will be better.

kou commented on lib/groonga-query-log/command/verify-server.rb:
```diff
@@ -230,6 +230,15 @@ def create_parser
           @options.rewrite_nullable_reference_number = boolean
         end
 
+        parser.on("--[no-]rewrite_regular_expression",
```
Ditto.

kou commented on lib/groonga-query-log/filter-rewriter.rb:
```diff
@@ -77,5 +80,21 @@ def rewrite_nullable_reference_number(filter)
+
+    def rewrite_regular_expression(filter)
+      filter.gsub(
+        /(?<before_expression>.+) *&& *(?<target_column>\w+) *@~ *\"\^\(\?\!\.\*(?<head_value>\w+) *(?:\| *\w+ *)*\)\.\+\$\"/
```
We can use `/.../x` for long regular expression:

```ruby
/(?<before_expression>.+)\ *&&
 \ *(?<target_column>\w+)\ *@~\ *
 \"\^\(\?\!\.\*(?<head_value>\w+)\ *(?:\|\ *\w+\ *)*\)\.\+\$\"/x
```

kou replied:
Is `(?<before_expression>.+)` needed?

kou replied:
`(?<target_column>\w+)` doesn't match `Table.index`.

kou replied:
`(?<target_column>\w+)` matches `column2あ`.
We're using `([a-zA-Z0-9_.]+)` for left hand side of expression in this class.
(The regular expression isn't enough because it matches `a...`.)

kou replied:
`\` is needless for `\"`.

kou replied:
`\^`should be `(?:\^|\\A)`.

kou replied:
`\$` should be `(?:\$|\\z)`.

kou replied:
`(?<head_value>\w+) *` doesn't match `xxx yyy`.

kou replied:
What about this?

```ruby
filter.gsub(/&& *(?<target_column>[a-zA-Z0-9_.]+) *@~ *"(?<pattern>.*?)"/) do |matched|
  target_column = $LAST_MATCH_INFO[:target_column]
  pattern = $LAST_MATCH_INFO[:pattern]

  case pattern
  when /\A(?<header>(?:\^|\\A)\(\?\!\.\*)
          (?<body>.+)
          (?<footer>\)\.[+*](?:\$|\\z))\z/x
    header = $LAST_MATCH_INFO[:header]
    body = $LAST_MATCH_INFO[:body]
    footer = $LAST_MATCH_INFO[:footer]
    conditions = body.split("|").collect do |word|
      "&! #{target_column} @ \"#{word.strip}\""
    end
    conditions.join(" ")
  else
    matched
  end
end
```

NOTE: Backslash escape in string literal isn't supported.

kou commented on lib/groonga-query-log/filter-rewriter.rb:
```diff
@@ -77,5 +80,21 @@ def rewrite_nullable_reference_number(filter)
+      filter.gsub(
+        /(?<before_expression>.+) *&& *(?<target_column>\w+) *@~ *\"\^\(\?\!\.\*(?<head_value>\w+) *(?:\| *\w+ *)*\)\.\+\$\"/
+      ) do |matched|
+        before_expression = $~[:before_expression]
```
Could you use `$LAST_MATCH_INFO` than `$~`?
`&~` isn't readable.

kou commented on test/test-filter-rewriter.rb:
```diff
@@ -110,4 +110,26 @@ def test_under_score
+
+    def test_no_rewrite
+      assert_equal("column1 @ \"value1\" && column2 @~ \"^(?!.*value2).+$\"",
+                   rewrite("column1 @ \"value1\" && column2 @~ \"^(?!.*value2).+$\"", false))
```
Is this needed?
Other options don't test this. Should we test this in all options?

komainu8 replied:
Surely, the purpose of this test that can correctly rewrite query or not so that this test is needless.
I'll fix it.

komainu8 commented:
@kou Thak you for your comments.
I've fixed them and added new tests by below commit.

* https://github.com/groonga/groonga-query-log/pull/41/commits/ef7c8c440c5da202db22cc9fcbd57f5d62f59f43
* https://github.com/groonga/groonga-query-log/pull/41/commits/7d2f36a9149ceb9f9641243f5ac68a66bd0111bf

kou commented on lib/groonga-query-log/filter-rewriter.rb:
```diff
@@ -77,5 +80,27 @@ def rewrite_nullable_reference_number(filter)
+        when /\A(?<header>(?:\^|\\A)\(\?\!\.\*)
+                (?<body>.+)
+                (?<footer>\)\.[+*](?:\$|\\z))\z/x
+          header = $LAST_MATCH_INFO[:header]
```
Could you remove this unused variable?

kou commented on lib/groonga-query-log/filter-rewriter.rb:
```diff
@@ -77,5 +80,27 @@ def rewrite_nullable_reference_number(filter)
+                (?<footer>\)\.[+*](?:\$|\\z))\z/x
+          header = $LAST_MATCH_INFO[:header]
+          body = $LAST_MATCH_INFO[:body]
+          footer = $LAST_MATCH_INFO[:footer]
```
ditto.

komainu8 commented:
@kou Thank you for your comments.
I've fixed them.

kou commented:
Thanks. Merged.
